File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char,int> mp1, mp2;
        vector<int>ans;
        int a  = s.size();
        int b = p.size();
        for(char c : p){
            mp1[c]++;
        }
        if(b > a) return ans;

        for(int i = 0; i < b; i++){
            mp2[s[i]]++;
        }
        if(mp1 == mp2) ans.push_back(0);
        for(int i = b; i < a; i++){
            mp2[s[i]]++;

            mp2[s[i- b]]--;
            if(mp2[s[i - b]] == 0){
                mp2.erase(s[i - b]);
            }
            if(mp1 == mp2) ans.push_back(i - b + 1);
        }
        return ans;
    }
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int n=s.length();
        int m=p.length();
        
        HashMap<Character,Integer>mp1 = new HashMap<>();
        HashMap<Character,Integer>mp2 = new HashMap<>();
        int j=0;
        ArrayList<Integer>result = new ArrayList<>();
        if(n<m){
            return result;
        }
        for(int i=0;i<m;i++){
            char ch = p.charAt(i);
            mp1.put(ch,mp1.getOrDefault(ch,0)+1);
        }
        for(int i=0;i<m;i++){
            char ch = s.charAt(i);
            mp2.put(ch,mp2.getOrDefault(ch,0)+1);
        }
        if(mp1.equals(mp2)){
             result.add(j);
        }
         for(int i=m;i<n;i++){
            char ch=s.charAt(j);
            mp2.put(ch, mp2.getOrDefault(ch,0)-1);
            if(mp2.get(ch)==0){
                mp2.remove(ch);
                
            }
            j++;
            mp2.put(s.charAt(i),mp2.getOrDefault(s.charAt(i),0)+1);

            if(mp1.equals(mp2)){
                result.add(j);
            }
        }
        return result;
    }
}